### PR TITLE
add machine state payload refinement

### PR DESCRIPTION
**Pull request by Kelerchian**

### CODE REVIEW

File: machine-runner/src/runner/runner.ts [hunk 1]
```diff
@@ -84,14 +88,34 @@ export type MachineRunner<SwarmProtocolName extends string, MachineName extends
    * }
    * machine.isDestroyed() // returns false
    */
-  noAutoDestroy: () => MachineRunnerIterableIterator<SwarmProtocolName, MachineName>
-} & MachineRunnerIterableIterator<SwarmProtocolName, MachineName>
+  noAutoDestroy: () => MachineRunnerIterableIterator<SwarmProtocolName, MachineName, StateUnion>
+
+  /**
+   * Add type refinement to the state payload produced by the machine-runner
```
Comment (rkuhn) [inline machine-runner/src/runner/runner.ts:94]: Correctly using this function requires passing ALL possible states, right? In that case, we should probably document this fact and check it at runtime (i.e. throw an exception if a state is missing).

File: machine-runner/src/runner/runner.ts [hunk 2]
```diff
@@ -402,15 +431,19 @@ export const createMachineRunnerInternal = <
...
   }
 
-  const defaultIterator: MachineRunnerIterableIterator<SwarmProtocolName, MachineName> =
+  const defaultIterator: MachineRunnerIterableIterator<SwarmProtocolName, MachineName, StateUnion> =
     MachineRunnerIterableIterator.make({
       events: emitter,
       inheritedDestruction: destruction,
     })
 
+  const refineStateType = (..._: Parameters<ThisMachineRunner['refineStateType']>) =>
```
Comment (rkuhn) [inline machine-runner/src/runner/runner.ts:440]: here we should check states list for completeness and throw if incomplete

File: machine-runner/tests/esm/runner.test.ts [hunk 3]
```diff
@@ -1150,6 +1151,28 @@ describe('typings', () => {
...
+
+    true as Expect<Equal<typeof stateOpaque['payload'], unknown>>
+    true as Expect<
+      Equal<
+        typeof refinedStateOpaque['payload'],
+        | StateFactory.PayloadOf<typeof ProtocolScorecard.Initial>
+        | StateFactory.PayloadOf<typeof ProtocolScorecard.Result>
+        | StateFactory.PayloadOf<typeof ProtocolScorecard.ScoreKeeping>
+      >
+    >
```
Comment (rkuhn) [inline machine-runner/tests/esm/runner.test.ts:1173]: :+1:

**Review by rkuhn — COMMENTED** [review]
Yes, looking good, just one more thing.

File: machine-runner/src/runner/runner.ts (hunk 1 above)
Comment (Kelerchian) [reply machine-runner/src/runner/runner.ts:94]: Currently, it doesn't. That's a good constraint thought, I'll add that.

**Pull request merged**